pyccel/ast/builtin_methods/set_methods.py

```diff
@@ -0,0 +1,76 @@
+# coding: utf-8
+#------------------------------------------------------------------------------------------#
+# This file is part of Pyccel which is released under MIT License. See the LICENSE file or #
+# go to https://github.com/pyccel/pyccel/blob/master/LICENSE for full license details.     #
+#------------------------------------------------------------------------------------------#
+"""
+The Set container has a number of built-in methods that are 
+always available.
+
+This module contains objects which describe these methods within Pyccel's AST.
+"""
+from pyccel.ast.datatypes import NativeVoid, NativeGeneric
+from pyccel.ast.internals import PyccelInternalFunction
+from pyccel.ast.builtins import PythonTuple
+
+__all__ = ('SetAdd',)
+
+class SetAdd(PyccelInternalFunction) :
+    """
+    Represents a call to the .add() method.
+    
+    Represents a call to the .add() method which add an element
+    to the set if it's not already present.
+    If the element is already in the set, the set remains unchanged.
+
+    Parameters
+    ----------
+    set_variable : TypedAstNode
+        The name of the set.
+
+    new_elem : TypedAstNode
+        The element that needs to be added to a set.
+    """
+    __slots__ = ("_set_variable", "_add_arg")
+    _attribute_nodes = ("_set_variable", "_add_arg")
+    _dtype = NativeVoid()
+    _shape = None
+    _order = None
+    _rank = 0
+    _precision = None
+    _class_type = NativeVoid()
+    name = 'add'
+
+    def __init__(self, set_variable, new_elem) -> None:
+        if (new_elem.rank > 0 and not isinstance(new_elem, PythonTuple)) :
+            raise TypeError("Can't add an unhashable type")
```

Comment:
As per the review summary message:
> It would be nice to add tests to ensure that a neat error is raised for inhomogeneous and non-hashable types.

If you have changed something so that the error is raised at the set instantiation please add a test for this